lsp-inline-completion.el

```diff
@@ -0,0 +1,487 @@
+  "Display the previous inline completion"
+  (interactive)
+  (unless (lsp-inline-completion--overlay-visible)
+    (error "Not showing suggestions"))
+  (setq lsp-inline-completion--current
+        (mod (1- lsp-inline-completion--current)
+             (length lsp-inline-completion--items)))
+
+  (lsp-inline-completion-show-overlay))
+
+;;;###autoload
+(defun lsp-inline-completion-display (&optional implicit)
+  "Displays the inline completions overlay"
+  (interactive)
+
+  (lsp--spinner-start)
+  (unwind-protect
+      (if-let* ((resp (lsp-request-while-no-input "textDocument/inlineCompletion"
+                                                  (lsp-inline-completion--params implicit)))
+                (items (lsp-inline-completion--parse-items resp)))
+
+          (progn
+            (lsp-inline-completion--clear-overlay)
+            (setq lsp-inline-completion--items items)
+            (setq lsp-inline-completion--current 0)
+            (setq lsp-inline-completion--start-point (point))
+            (lsp-inline-completion-show-overlay))
+        (unless implicit
+          (message "No Suggestions!")))
+    ;; Clean up
+    (lsp--spinner-stop)))
+
+
+;; Inline Completion Mode
+(defcustom lsp-inline-completion-enable t
+  "If non-nil it will enable inline completions on idle."
+  :type 'boolean
+  :group 'lsp-mode
+  :package-version '(lsp-mode . "9.0.1"))
+
+(defcustom lsp-inline-completion-idle-delay 2
+  "The number of seconds before trying to fetch inline completions, when
+lsp-inline-completion-mode is active"
+  :type 'number
+  :group 'lsp-mode
+  :package-version '(lsp-mode . "9.0.1"))
+
+(defcustom lsp-inline-completion-inhibit-predicates nil
+  "When a function of this list returns non nil, lsp-inline-completion-mode will not show the completion"
+  :type '(repeat function)
+  :group 'lsp-mode)
+
+(defvar-local lsp-inline-completion--idle-timer nil
+  "The idle timer used by lsp-inline-completion-mode")
+
+(define-minor-mode lsp-inline-completion-mode
+  "Mode automatically displaying inline completions."
+  :lighter nil
+  (cond
+   ((and lsp-inline-completion-mode lsp--buffer-workspaces)
+    (add-hook 'lsp-on-change-hook #'lsp-inline-completion--after-change nil t))
+   (t
+    (when lsp-inline-completion--idle-timer
+      (cancel-timer lsp-inline-completion--idle-timer))
+
+    (lsp-inline-completion-cancel)
+
+    (remove-hook 'lsp-on-change-hook #'lsp-inline-completion--after-change t))))
+
+(defun lsp-inline-completion--maybe-display (buffer)
+  (when (and (buffer-live-p buffer)
+             (eq (current-buffer) buffer)
+             (--none? (funcall it) lsp-inline-completion-inhibit-predicates))
+    (setq last-command this-command)
+    (setq this-command 'lsp-inline-completion-display)
+    (lsp-inline-completion-display 'implicit)))
+
+(defun lsp-inline-completion--after-change (&rest _)
+  (when (and lsp-inline-completion-mode lsp--buffer-workspaces)
+    (when lsp-inline-completion--idle-timer
```

Comment:
perhaps we can move this outside of the `when` check to make sure `lsp-inline-completion--idle-timer` always be cancelled?